cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1695,6 +1692,389 @@ struct GroupedMinMaxFactory {
+    has_values_ = TypedBufferBuilder<bool>(ctx->memory_pool());
+    has_nulls_ = TypedBufferBuilder<bool>(ctx->memory_pool());
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    auto added_groups = new_num_groups - num_groups_;
+    DCHECK_GE(added_groups, 0);
+    num_groups_ = new_num_groups;
+    firsts_.resize(new_num_groups);
+    lasts_.resize(new_num_groups);
+    RETURN_NOT_OK(has_values_.Append(added_groups, false));
+    RETURN_NOT_OK(has_nulls_.Append(added_groups, false));
+    return Status::OK();
+  }
+
+  Status Consume(const ExecSpan& batch) override {
+    return VisitGroupedValues<Type>(
+        batch,
+        [&](uint32_t g, std::string_view val) {
+          if (!firsts_[g]) {
+            firsts_[g].emplace(val.data(), val.size(), allocator_);
+          }
+          lasts_[g].emplace(val.data(), val.size(), allocator_);
+          bit_util::SetBit(has_values_.mutable_data(), g);
+          return Status::OK();
+        },
+        [&](uint32_t g) {
+          bit_util::SetBit(has_nulls_.mutable_data(), g);
+          return Status::OK();
+        });
+  }
+
+  Status Merge(GroupedAggregator&& raw_other,
+               const ArrayData& group_id_mapping) override {
+    auto other = checked_cast<GroupedFirstLastImpl*>(&raw_other);
+    auto g = group_id_mapping.GetValues<uint32_t>(1);
+    for (uint32_t other_g = 0; static_cast<int64_t>(other_g) < group_id_mapping.length;
+         ++other_g, ++g) {
+      if (!firsts_[*g]) {
+        firsts_[*g] = std::move(other->firsts_[other_g]);
+      }
+      lasts_[*g] = std::move(other->lasts_[other_g]);
+
+      if (bit_util::GetBit(other->has_values_.data(), other_g)) {
+        bit_util::SetBit(has_values_.mutable_data(), *g);
+      }
+      if (bit_util::GetBit(other->has_nulls_.data(), other_g)) {
+        bit_util::SetBit(has_nulls_.mutable_data(), *g);
+      }
+    }
+    return Status::OK();
+  }
+
+  Result<Datum> Finalize() override {
+    // aggregation for group is valid if there was at least one value in that group
+    ARROW_ASSIGN_OR_RAISE(auto null_bitmap, has_values_.Finish());
+
+    if (!options_.skip_nulls) {
+      // ... and there were no nulls in that group
+      ARROW_ASSIGN_OR_RAISE(auto has_nulls, has_nulls_.Finish());
+      arrow::internal::BitmapAndNot(null_bitmap->data(), 0, has_nulls->data(), 0,
+                                    num_groups_, 0, null_bitmap->mutable_data());
+    }
+
+    auto firsts = ArrayData::Make(type_, num_groups_, {null_bitmap, nullptr});
+    auto lasts = ArrayData::Make(type_, num_groups_, {std::move(null_bitmap), nullptr});
+    RETURN_NOT_OK(MakeOffsetsValues(firsts.get(), firsts_));
+    RETURN_NOT_OK(MakeOffsetsValues(lasts.get(), lasts_));
+    return ArrayData::Make(out_type(), num_groups_, {nullptr},
+                           {std::move(firsts), std::move(lasts)});
+  }
+
+  template <typename T = Type>
+  enable_if_base_binary<T, Status> MakeOffsetsValues(
```

Comment:
This is copied from the binary version of the grouped min/max kernel. I will try to refactor this to a base class.

Reply:
I failed to address this because the templates make it hard to refactor and share the logic between "MinMax" and "FirstLast" for binary types. if that's ok I would like to leave that as follow up to better refactor this and avoid adding complexity to this PR.